File: go/internal/csvdiff/options.go
package csvdiff

import (
	"fmt"
	"slices"
	"strings"
)

// Engine names the comparison backend.
//
// EngineAuto resolves to the first concrete engine that can run here, in
// declaration order. Every concrete engine must return identical Counts and
// Columns for the same input; the tests and CI both assert that.
type Engine string

const (
	// EngineAuto picks the first available concrete engine.
	EngineAuto Engine = "auto"
	// EngineDuckDB runs the comparison in DuckDB: out-of-core, handles files larger than RAM.
	EngineDuckDB Engine = "duckdb"
	// EngineSortMerge sorts both files by key and walks them together: bounded
	// memory, spills to disk, size limited by disk rather than RAM.
	EngineSortMerge Engine = "sortmerge"
	// EngineNative parses with encoding/csv and joins in a map. In-memory only.
	EngineNative Engine = "native"
)

// ConcreteEngines lists the real backends in EngineAuto preference order.
var ConcreteEngines = []Engine{EngineDuckDB, EngineSortMerge, EngineNative}

// ParseEngine validates a command-line spelling.
func ParseEngine(s string) (Engine, error) {
	e := Engine(strings.ToLower(strings.TrimSpace(s)))
	if e == EngineAuto || slices.Contains(ConcreteEngines, e) {
		return e, nil
	}
	valid := []string{string(EngineAuto)}
	for _, c := range ConcreteEngines {
		valid = append(valid, string(c))
	}
	return "", fmt.Errorf("unknown engine: %s. Choose one of %s", s, strings.Join(valid, ", "))
}

// DefaultMaxRows is how many rows each report section embeds by default.
// Counts are always exact regardless.
const DefaultMaxRows = 50_000

// Options is everything that varies per comparison. Nothing about a specific
// dataset belongs in the code, so key columns, compared columns and
// normalisation are all runtime parameters.
//
// The JSON names are snake_case to match the other implementations, which read
// and write the same contract.
type Options struct {
	// Key is the composite key; at least one column is required.
	Key []string `json:"-"`
	// Compare lists the columns to diff, or nil for every common non-key column.
	Compare []string `json:"-"`
	// Ignore lists columns to skip entirely.
	Ignore []string `json:"-"`

	Trim        bool    `json:"trim"`
	IgnoreCase  bool    `json:"ignore_case"`
	EmptyIsNull bool    `json:"empty_is_null"`
	Tolerance   float64 `json:"tolerance"`
	MaxRows     int     `json:"max_rows"`
	Delimiter   string  `json:"delimiter"`
	Encoding    string  `json:"encoding"`
	EngineName  string  `json:"engine"`
	Threads     int     `json:"threads"`
	MemoryLimit string  `json:"memory_limit"`
	ExportDir   string  `json:"export_dir"`
}

// NewOptions returns the defaults, which callers then override.
func NewOptions() Options {
	return Options{
		MaxRows:    DefaultMaxRows,
		Encoding:   "utf-8",
		EngineName: string(EngineAuto),
	}
}

// Validate checks the options are usable and normalises the defaults.
func (o *Options) Validate() error {
	if len(o.Key) == 0 {
		return fmt.Errorf("at least one key column is required")
	}
	if o.Tolerance < 0 {
		return fmt.Errorf("--tolerance must not be negative, got %v", o.Tolerance)
	}
	if o.MaxRows <= 0 {
		return fmt.Errorf("--max-rows must be positive, got %d", o.MaxRows)
	}
	if o.Threads < 0 {
		return fmt.Errorf("--threads must not be negative, got %d", o.Threads)
	}
	if o.Encoding == "" {
		o.Encoding = "utf-8"
	}
	if o.EngineName == "" {
		o.EngineName = string(EngineAuto)
	}
	if _, err := ParseEngine(o.EngineName); err != nil {
		return err
	}
	if o.Delimiter != "" && len([]rune(o.Delimiter)) != 1 {
		return fmt.Errorf("--delimiter must be a single character, got %q", o.Delimiter)
	}
	return nil
}

// DelimiterRune returns the forced delimiter, or 0 when it should be sniffed.
func (o Options) DelimiterRune() rune {
	if o.Delimiter == "" {
		return 0
	}
	return []rune(o.Delimiter)[0]
}

// ParseList splits "a, b ,c" into a slice; an empty string yields nil.
func ParseList(s string) []string {
	if strings.TrimSpace(s) == "" {
		return nil
	}
	parts := strings.Split(s, ",")
	out := make([]string, 0, len(parts))
	for _, p := range parts {
		if v := strings.TrimSpace(p); v != "" {
			out = append(out, v)
		}
	}
	return out
}

File: rust/src/options.rs
//! Runtime parameters and engine names.

use serde::Serialize;
use std::fmt;
use std::str::FromStr;

use crate::error::{Error, Result};

/// Which comparison backend to run.
///
/// [`Engine::Auto`] resolves to the first concrete engine that can run here, in
/// declaration order. Every concrete engine must return identical `counts` and
/// `columns` for the same input; the tests and CI both assert that.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Engine {
    /// Picks the first available concrete engine.
    Auto,
    /// DuckDB: out-of-core, handles files larger than RAM.
    DuckDb,
    /// Polars: in-memory, columnar, multi-threaded.
    Polars,
    /// Byte-level: maps the file and indexes it in place, building no string
    /// for a cell unless that cell reaches the report.
    Turbo,
    /// External sort then merge join: bounded memory, spills to disk, size
    /// limited by disk rather than RAM.
    SortMerge,
    /// This project, over the `csv` crate: in-memory, row-oriented.
    Native,
}

impl Engine {
    /// The real backends, in `Auto` preference order.
    pub const CONCRETE: [Engine; 4] = [
        Engine::DuckDb,
        Engine::Polars,
        Engine::SortMerge,
        Engine::Native,
    ];

    /// The name used on the command line and in the report.
    pub fn label(self) -> &'static str {
        match self {
            Engine::Auto => "auto",
            Engine::DuckDb => "duckdb",
            Engine::Polars => "polars",
            Engine::Turbo => "turbo",
            Engine::SortMerge => "sortmerge",
            Engine::Native => "native",
        }
    }
}

impl FromStr for Engine {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self> {
        match s.trim().to_ascii_lowercase().as_str() {
            "auto" => Ok(Engine::Auto),
            "duckdb" => Ok(Engine::DuckDb),
            "polars" => Ok(Engine::Polars),
            "turbo" => Ok(Engine::Turbo),
            "sortmerge" => Ok(Engine::SortMerge),
            "native" => Ok(Engine::Native),
            other => Err(Error::new(format!(
                "unknown engine: {other}. Choose one of auto, duckdb, polars, turbo, sortmerge, native"
            ))),
        }
    }
}

impl fmt::Display for Engine {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.label())
    }
}

/// How many rows each report section embeds by default. Counts are always exact
/// regardless.
pub const DEFAULT_MAX_ROWS: usize = 50_000;

/// Everything that varies per comparison. Nothing about a specific dataset
/// belongs in the code, so key columns, compared columns and normalisation are
/// all runtime parameters.
///
/// The JSON names are snake_case to match the other implementations, which read
/// and write the same contract.
#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct Options {
    /// The composite key; at least one column is required.
    #[serde(skip)]
    pub key: Vec<String>,
    /// The columns to diff, or `None` for every common non-key column.
    #[serde(skip)]
    pub compare: Option<Vec<String>>,
    /// Columns to skip entirely.
    #[serde(skip)]
    pub ignore: Vec<String>,

    pub trim: bool,
    pub ignore_case: bool,
    pub empty_is_null: bool,
    pub tolerance: f64,
    pub max_rows: usize,
    pub delimiter: Option<char>,
    pub encoding: String,
    pub engine: String,
    pub threads: Option<usize>,
    pub memory_limit: Option<String>,
    pub export_dir: Option<String>,
}

impl Default for Options {
    fn default() -> Self {
        Options {
            key: Vec::new(),
            compare: None,
            ignore: Vec::new(),
            trim: false,
            ignore_case: false,
            empty_is_null: false,
            tolerance: 0.0,
            max_rows: DEFAULT_MAX_ROWS,
            delimiter: None,
            encoding: "utf-8".to_string(),
            engine: Engine::Auto.label().to_string(),
            threads: None,
            memory_limit: None,
            export_dir: None,
        }
    }
}

impl Options {
    /// Options for a comparison on the given key, everything else defaulted.
    pub fn with_key<I, S>(key: I) -> Self
    where
        I: IntoIterator<Item = S>,
        S: Into<String>,
    {
        Options {
            key: key.into_iter().map(Into::into).collect(),
            ..Options::default()
        }
    }

    /// Checks the options are usable and normalises the defaults.
    pub fn validate(&mut self) -> Result<()> {
        if self.key.is_empty() {
            return Err(Error::new("at least one key column is required"));
        }
        if self.tolerance < 0.0 {
            return Err(Error::new(format!(
                "--tolerance must not be negative, got {}",
                self.tolerance
            )));
        }
        if self.max_rows == 0 {
            return Err(Error::new("--max-rows must be positive, got 0"));
        }
        if self.threads == Some(0) {
            return Err(Error::new("--threads must be positive, got 0"));
        }
        if self.encoding.is_empty() {
            self.encoding = "utf-8".to_string();
        }
        if self.engine.is_empty() {
            self.engine = Engine::Auto.label().to_string();
        }
        self.engine_name()?;
        Ok(())
    }

    /// The requested engine, parsed.
    pub fn engine_name(&self) -> Result<Engine> {
        self.engine.parse()
    }

    /// The forced delimiter as a byte, or `None` when it should be sniffed.
    ///
    /// The CSV readers here are byte-oriented, so a multi-byte delimiter cannot
    /// be honoured; that is rejected rather than silently truncated.
    pub fn delimiter_byte(&self) -> Result<Option<u8>> {
        match self.delimiter {
            None => Ok(None),
            Some(c) if c.is_ascii() => Ok(Some(c as u8)),
            Some(c) => Err(Error::new(format!(
                "--delimiter must be an ASCII character, got {c:?}"
            ))),
        }
    }
}

/// Splits `"a, b ,c"` into a list; an empty string yields nothing at all.
pub fn parse_list(value: &str) -> Vec<String> {
    value
        .split(',')
        .map(str::trim)
        .filter(|s| !s.is_empty())
        .map(str::to_string)
        .collect()
}
